### ISSUE TITLE
Test ACC 0.0%?

### ISSUE THREAD

**Issue by davidsonic** [rxtan2/Learning-Similarity-Conditions#4]
Hi, Is it normal to to get printed test acc 0? During training, the printed log is like this:
Test set: Average loss: 0.0767, Accuracy: 0.00%                                                      
                                                                                                                                
Train Epoch: 11 [0/20000]       Loss: 0.0103 (0.0103)   Acc: 100.00% (100.00%)  Emb_Norm: 0.37 (0.37)
Train Epoch: 11 [3200/20000]    Loss: 0.0031 (0.0145)   Acc: 100.00% (0.00%)    Emb_Norm: 0.35 (0.35)
Train Epoch: 11 [6400/20000]    Loss: 0.0148 (0.0149)   Acc: 0.00% (0.00%)      Emb_Norm: 0.32 (0.35)
Train Epoch: 11 [9600/20000]    Loss: 0.0228 (0.0153)   Acc: 0.00% (0.00%)      Emb_Norm: 0.38 (0.35)
Train Epoch: 11 [12800/20000]   Loss: 0.0302 (0.0160)   Acc: 0.00% (0.00%)      Emb_Norm: 0.35 (0.35)
Train Epoch: 11 [16000/20000]   Loss: 0.0158 (0.0158)   Acc: 0.00% (0.00%)      Emb_Norm: 0.35 (0.35)
Train Epoch: 11 [19200/20000]   Loss: 0.0017 (0.0158)   Acc: 100.00% (0.00%)    Emb_Norm: 0.36 (0.35)
                                                                              
Test set: Average loss: 0.0782, Accuracy: 0.00%                                                      
                                                                                                      
Train Epoch: 12 [0/20000]       Loss: 0.0078 (0.0078)   Acc: 100.00% (100.00%)  Emb_Norm: 0.40 (0.40)
Train Epoch: 12 [3200/20000]    Loss: 0.0032 (0.0117)   Acc: 100.00% (0.00%)    Emb_Norm: 0.37 (0.36)                      
Train Epoch: 12 [6400/20000]    Loss: 0.0000 (0.0128)   Acc: 100.00% (0.00%)    Emb_Norm: 0.34 (0.35)
Train Epoch: 12 [9600/20000]    Loss: 0.0000 (0.0123)   Acc: 100.00% (0.00%)    Emb_Norm: 0.35 (0.35)
Train Epoch: 12 [12800/20000]   Loss: 0.0000 (0.0123)   Acc: 100.00% (0.00%)    Emb_Norm: 0.35 (0.35)
Train Epoch: 12 [16000/20000]   Loss: 0.0502 (0.0129)   Acc: 0.00% (0.00%)      Emb_Norm: 0.32 (0.35)
Train Epoch: 12 [19200/20000]   Loss: 0.0018 (0.0137)   Acc: 100.00% (0.00%)    Emb_Norm: 0.36 (0.35)
                                                                                                     
Test set: Average loss: 0.0742, Accuracy: 0.00%                                                      
                                                                                                      
Train Epoch: 13 [0/20000]       Loss: 0.0089 (0.0089)   Acc: 0.00% (0.00%)      Emb_Norm: 0.37 (0.37)
Train Epoch: 13 [3200/20000]    Loss: 0.0018 (0.0096)   Acc: 100.00% (0.00%)    Emb_Norm: 0.37 (0.35)
Train Epoch: 13 [6400/20000]    Loss: 0.0150 (0.0107)   Acc: 0.00% (0.00%)      Emb_Norm: 0.35 (0.34)                           
Train Epoch: 13 [9600/20000]    Loss: 0.0059 (0.0107)   Acc: 100.00% (0.00%)    Emb_Norm: 0.35 (0.35)
Train Epoch: 13 [12800/20000]   Loss: 0.0186 (0.0103)   Acc: 0.00% (0.00%)      Emb_Norm: 0.33 (0.35)                                                                                                      
Train Epoch: 13 [16000/20000]   Loss: 0.0000 (0.0110)   Acc: 100.00% (0.00%)    Emb_Norm: 0.34 (0.34)
Train Epoch: 13 [19200/20000]   Loss: 0.0214 (0.0110)   Acc: 0.00% (0.00%)      Emb_Norm: 0.33 (0.35)
                                                                                                     
Test set: Average loss: 0.0718, Accuracy: 0.00%                                                      
                                                                                                     
Train Epoch: 14 [0/20000]       Loss: 0.0041 (0.0041)   Acc: 100.00% (100.00%)  Emb_Norm: 0.36 (0.36)
Train Epoch: 14 [3200/20000]    Loss: 0.0004 (0.0118)   Acc: 100.00% (0.00%)    Emb_Norm: 0.35 (0.35) 
Train Epoch: 14 [6400/20000]    Loss: 0.0157 (0.0108)   Acc: 0.00% (0.00%)      Emb_Norm: 0.33 (0.34)
Train Epoch: 14 [9600/20000]    Loss: 0.0367 (0.0112)   Acc: 0.00% (0.00%)      Emb_Norm: 0.33 (0.34)
Train Epoch: 14 [12800/20000]   Loss: 0.0073 (0.0109)   Acc: 0.00% (0.00%)      Emb_Norm: 0.31 (0.34)                           
Train Epoch: 14 [16000/20000]   Loss: 0.0208 (0.0108)   Acc: 0.00% (0.00%)      Emb_Norm: 0.33 (0.34)
Train Epoch: 14 [19200/20000]   Loss: 0.0098 (0.0107)   Acc: 100.00% (0.00%)    Emb_Norm: 0.38 (0.34)                                                                                                      
                                                                               
Test set: Average loss: 0.0740, Accuracy: 0.00%               
                                                              
Train Epoch: 15 [0/20000]       Loss: 0.0000 (0.0000)   Acc: 100.00% (100.00%)  Emb_Norm: 0.33 (0.33)
Train Epoch: 15 [3200/20000]    Loss: 0.0009 (0.0093)   Acc: 100.00% (0.00%)    Emb_Norm: 0.37 (0.35)
Train Epoch: 15 [6400/20000]    Loss: 0.0000 (0.0089)   Acc: 100.00% (0.00%)    Emb_Norm: 0.34 (0.34)
Train Epoch: 15 [9600/20000]    Loss: 0.0220 (0.0084)   Acc: 100.00% (0.00%)    Emb_Norm: 0.32 (0.34)
Train Epoch: 15 [12800/20000]   Loss: 0.0000 (0.0091)   Acc: 100.00% (0.00%)    Emb_Norm: 0.35 (0.34)
Train Epoch: 15 [16000/20000]   Loss: 0.0204 (0.0095)   Acc: 0.00% (0.00%)      Emb_Norm: 0.30 (0.34)
Train Epoch: 15 [19200/20000]   Loss: 0.0220 (0.0095)   Acc: 0.00% (0.00%)      Emb_Norm: 0.33 (0.34)

Test set: Average loss: 0.0727, Accuracy: 0.00%

During testing the printed information is:
Loading Train Dataset                                                                              
Loading Test Dataset                                       
Loading Val Dataset                                                            
  + Number of params: 2805796                                 
/media/sdb/jiali/env/lib/python3.6/site-packages/torch/nn/modules/container.py:92: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
  input = module(input)                                                      
                                                              
Test set: Average loss: 0.0739, Accuracy: 0.00%

**Comment by rxtan2** [comment]
Hi, that's not an expected output.  Is the accuracy on the validation always 0 even after the first epoch?  Are you using the specified version of pytorch?

**Comment by davidsonic** [comment]
OK, now it seems to work. 
Test set: Average loss: 0.1815, Accuracy: 59.49%
initial test acc: 0.594875
Train Epoch: 1 [0/20000]        Loss: 0.1798 (0.1798)   Acc: 56.25% (56.25%)    Emb_Norm: 0.75 (0.75)
Train Epoch: 1 [3200/20000]     Loss: 0.1320 (0.1557)   Acc: 75.00% (65.38%)    Emb_Norm: 0.64 (0.68)
Train Epoch: 1 [6400/20000]     Loss: 0.1224 (0.1489)   Acc: 78.12% (67.18%)    Emb_Norm: 0.62 (0.65)
Train Epoch: 1 [9600/20000]     Loss: 0.1213 (0.1428)   Acc: 68.75% (68.68%)    Emb_Norm: 0.56 (0.63)
Train Epoch: 1 [12800/20000]    Loss: 0.1146 (0.1376)   Acc: 65.62% (69.84%)    Emb_Norm: 0.54 (0.61)
Train Epoch: 1 [16000/20000]    Loss: 0.0925 (0.1334)   Acc: 78.12% (70.82%)    Emb_Norm: 0.49 (0.59)
Train Epoch: 1 [19200/20000]    Loss: 0.0692 (0.1291)   Acc: 87.50% (71.96%)    Emb_Norm: 0.43 (0.57)
Test set: Average loss: 0.1226, Accuracy: 73.63%
validation acc: 0.736325

**Comment by rxtan2** [comment]
Nice! What was the error that caused the accuracy to be 0?

**Comment by davidsonic** [comment]
Hi, the error will go away by modifying the line in accuracy_id function to 
((pred > 0)*(c.cpu().data.numpy() == c_id)).sum()*1.0/(c.cpu().data.numpy() == c_id).sum()
and changing pytorch 0.x APIs to compatible pytorch 1.x APIs when necessary.  Issue closed.

**Comment by davidsonic** [comment]
There still seems a big gap between evaluation accuracy and training acc, is the final test performance 86% expected?

Test set: Average loss: 0.0701, Accuracy: 85.44%

Train Epoch: 13 [0/20000]       Loss: 0.0012 (0.0012)   Acc: 100.00% (100.00%)  Emb_Norm: 0.37 (0.37)
Train Epoch: 13 [3200/20000]    Loss: 0.0014 (0.0096)   Acc: 100.00% (98.92%)   Emb_Norm: 0.32 (0.35)
Train Epoch: 13 [6400/20000]    Loss: 0.0262 (0.0107)   Acc: 96.88% (98.65%)    Emb_Norm: 0.37 (0.35)
Train Epoch: 13 [9600/20000]    Loss: 0.0067 (0.0110)   Acc: 100.00% (98.62%)   Emb_Norm: 0.35 (0.35)
Train Epoch: 13 [12800/20000]   Loss: 0.0095 (0.0113)   Acc: 100.00% (98.57%)   Emb_Norm: 0.33 (0.35)
Train Epoch: 13 [16000/20000]   Loss: 0.0048 (0.0117)   Acc: 100.00% (98.56%)   Emb_Norm: 0.37 (0.35)
Train Epoch: 13 [19200/20000]   Loss: 0.0084 (0.0115)   Acc: 100.00% (98.65%)   Emb_Norm: 0.33 (0.35)

Test set: Average loss: 0.0711, Accuracy: 85.72%

Train Epoch: 14 [0/20000]       Loss: 0.0055 (0.0055)   Acc: 100.00% (100.00%)  Emb_Norm: 0.37 (0.37)
Train Epoch: 14 [3200/20000]    Loss: 0.0049 (0.0095)   Acc: 100.00% (99.04%)   Emb_Norm: 0.31 (0.35)
Train Epoch: 14 [6400/20000]    Loss: 0.0120 (0.0097)   Acc: 96.88% (98.87%)    Emb_Norm: 0.36 (0.35)
Train Epoch: 14 [9600/20000]    Loss: 0.0000 (0.0101)   Acc: 100.00% (98.83%)   Emb_Norm: 0.33 (0.34)
Train Epoch: 14 [12800/20000]   Loss: 0.0024 (0.0098)   Acc: 100.00% (98.89%)   Emb_Norm: 0.36 (0.34)
Train Epoch: 14 [16000/20000]   Loss: 0.0000 (0.0098)   Acc: 100.00% (98.90%)   Emb_Norm: 0.37 (0.34)
Train Epoch: 14 [19200/20000]   Loss: 0.0064 (0.0102)   Acc: 100.00% (98.84%)   Emb_Norm: 0.34 (0.34)

Test set: Average loss: 0.0714, Accuracy: 85.72%

Train Epoch: 15 [0/20000]       Loss: 0.0000 (0.0000)   Acc: 100.00% (100.00%)  Emb_Norm: 0.35 (0.35)
Train Epoch: 15 [3200/20000]    Loss: 0.0101 (0.0071)   Acc: 100.00% (99.44%)   Emb_Norm: 0.32 (0.34)
Train Epoch: 15 [6400/20000]    Loss: 0.0043 (0.0086)   Acc: 100.00% (99.21%)   Emb_Norm: 0.35 (0.34)
Train Epoch: 15 [9600/20000]    Loss: 0.0017 (0.0082)   Acc: 100.00% (99.22%)   Emb_Norm: 0.34 (0.34)
Train Epoch: 15 [12800/20000]   Loss: 0.0000 (0.0086)   Acc: 100.00% (99.14%)   Emb_Norm: 0.34 (0.34)
Train Epoch: 15 [16000/20000]   Loss: 0.0050 (0.0088)   Acc: 100.00% (99.13%)   Emb_Norm: 0.34 (0.34)
Train Epoch: 15 [19200/20000]   Loss: 0.0029 (0.0091)   Acc: 100.00% (99.08%)   Emb_Norm: 0.32 (0.34)

Test set: Average loss: 0.0693, Accuracy: 85.95%


Test set: Average loss: 0.0689, Accuracy: 86.05%
